plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java

```diff
@@ -0,0 +1,707 @@
+    private final Faker faker;
+    private final FakerTableHandle table;
+    private final List<BiConsumer<BlockBuilder, Long>> generators;
+    private long completedBytes;
+    private long completedRows;
+
+    private final PageBuilder pageBuilder;
+
+    private boolean closed;
+
+    FakerPageSource(Faker faker, Random random, List<FakerColumnHandle> columns, FakerTableHandle table)
+    {
+        this.faker = requireNonNull(faker, "faker is null");
+        this.random = requireNonNull(random, "random is null");
+        this.table = requireNonNull(table, "table is null");
+        List<Type> types = requireNonNull(columns, "columns is null")
+                .stream()
+                .map(FakerColumnHandle::type)
+                .collect(toList());
+        this.generators = columns
+                .stream()
+                .map(column -> constraintedValueGenerator(
+                        column,
+                        table.constraint().getDomains().get().getOrDefault(column, Domain.all(column.type()))))
+                .collect(toList());
+        // TODO create generators for every column
+        this.pageBuilder = new PageBuilder(types);
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed && pageBuilder.isEmpty();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (!closed) {
+            int positions = (int) Math.min(table.limit() - completedRows, ROWS_PER_REQUEST);
+            if (positions <= 0) {
+                closed = true;
+            }
+            else {
+                pageBuilder.declarePositions(positions);
+                for (int column = 0; column < generators.size(); column++) {
+                    BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(column);
+                    BiConsumer<BlockBuilder, Long> generator = generators.get(column);
+                    for (int i = 0; i < positions; i++) {
+                        generator.accept(blockBuilder, completedRows + i);
+                    }
+                }
+                completedRows += positions;
+            }
+        }
+
+        // only return a page if the buffer is full, or we are finishing
+        if ((closed && !pageBuilder.isEmpty()) || pageBuilder.isFull()) {
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            completedBytes += page.getSizeInBytes();
+            return page;
+        }
+
+        return null;
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return pageBuilder.getSizeInBytes();
+    }
+
+    @Override
+    public void close()
+    {
+        closed = true;
+    }
+
+    private BiConsumer<BlockBuilder, Long> constraintedValueGenerator(FakerColumnHandle handle, Domain domain)
```

Comment:
Please make an explicit interface instead of `BiConsumer<BlockBuilder, Long>`.
That makes it easier to read the code and can you can track usages of it from IDE